user_tools/src/spark_rapids_tools/tools/qualification_stats_report.py

```diff
@@ -59,36 +60,85 @@ def _read_csv_files(self) -> None:
+        self.execs_df.rename(columns={'Exec Name': 'Operator'}, inplace=True)
+        self.unsupported_operators_df.rename(columns={'Unsupported Operator': 'Operator'},
+                                             inplace=True)
+        self.logger.info('Preprocessing dataframes completed.')
+
     def _merge_dataframes(self) -> None:
         self.logger.info('Merging dataframes to get stats...')
-        # Merge unsupported_operators_df with stages_df on App ID and Stage ID
-        merged_df = pd.merge(self.unsupported_operators_df, self.stages_df,
-                             on=['App ID', 'Stage ID'])
-
-        agg_unsupported_df = (merged_df.groupby(['App ID', 'SQL ID', 'Unsupported Operator']).agg(
-            Count=('Unsupported Operator', 'size'),
-            Impacted_Stage_Duration=('Stage Duration', 'sum'),
-            App_Duration=('App Duration', 'first'),
-            Stage_Task_Duration=('Stage Task Duration', 'sum')
-        ).reset_index())
-
-        agg_unsupported_df['% of Stage Duration'] = (
-                (agg_unsupported_df['Impacted_Stage_Duration'] /
-                 agg_unsupported_df['App_Duration']) * 100).round(3)
-
-        agg_unsupported_df['Supported'] = False
-        final_df = agg_unsupported_df.rename(columns={
-            'Unsupported Operator': 'Operator',
-            'Impacted_Stage_Duration': 'Impacted Stage duration(seconds)',
-            'Stage_Task_Duration': 'Stage Task Exec duration(seconds)'
-        })
+        self._preprocess_dataframes()
+
+        # Merge execs_df with stages_df
+        merged_df = self.execs_df.merge(self.stages_df, left_on=['App ID', 'Exec Stages'],
+                                        right_on=['App ID', 'Stage ID'], how='left')
+
+        # Merge with unsupported_operators_df to find unsupported operations
+        merged_df = merged_df.merge(self.unsupported_operators_df,
+                                    on=['App ID', 'SQL ID', 'Stage ID', 'Operator'],
+                                    how='left', indicator=True)
+        merged_df['Supported'] = merged_df['_merge'] == 'left_only'
+        merged_df.drop(columns=['_merge', 'Exec Stages'], inplace=True)
+
+        # Calculate total duration by summing unique stages per SQLID
+        total_duration_df = merged_df.drop_duplicates(subset=['App ID', 'SQL ID', 'Stage ID']) \
+            .groupby(['App ID', 'SQL ID'])['StageTaskDuration'] \
+            .sum().reset_index().rename(columns={'StageTaskDuration': 'TotalSQLDuration'})
+
+        merged_df = merged_df.merge(total_duration_df, on=['App ID', 'SQL ID'], how='left')
+
+        # Mark unique stage task durations
+        merged_df['Unique StageTaskDuration'] = ~merged_df.duplicated(
+            ['App ID', 'SQL ID', 'Operator', 'Stage ID', 'Supported'])
+        merged_df['Adjusted StageTaskDuration'] = (merged_df['StageTaskDuration'] *
+                                                   merged_df['Unique StageTaskDuration'])
```

Comment:
QQ: what is the meaning to multiply `StageTaskDuration` and `Unique StageTaskDuration`?

Reply:
`StageTaskDuration` - is the duration of the StageTask(CPU core) duration of  a particular operator in a specific stage.
Unique StageTaskDuration: Flags the first occurrence of each unique stage task duration to ensure correct summation to avoid double counting. 
For example:
```
>>> merged_df
   AppId  SQLID  Operator Name     Stage ID  StageTaskDuration  Supported  TotalSQLDuration  Unique StageTaskDuration  Adjusted StageTaskDuration
0      1      1           Sort           5                 50       True               230           True                          50
1      1      1         Filter           5                 50       True               230           True                          50
2      1      1        Project           5                 50       True               230           True                          50
3      1      1        Project           5                 50       True               230          False                           0
4      1      1         Filter           4                 20       True               230           True                          20

```
In the above df, Project is present twice for StageID=5.  Since StageTaskDuration  is total time taken for that stage, we need to consider it only once i.e 50 should not be double counted. But we need to both Project's to get the count of Execs for that SQLID. If we filter it earlier itself, then count of Project will be incorrect.

Reply:
Thanks for this explanation